doc/inband-telemetry/SAI-Proposal-Advanced-Telemetry-draft.md

```diff
@@ -0,0 +1,421 @@
+ SAI-Version | XXX
+ 
+-------------------------------------------------------------------------------
+
+# Overview #
+
+This draft describes the API proposal to enable advanced telemetry capabilities in a network device. The main goal of this advanced telemetry proposal is to achieve per-packet network visibility with low overhead. The network device should be able to inspect and take actions on each individual data packet.
+
+## Telemetry data plane workflow ##
+
+![workflow](figures/workflow.png "Figure 1: Telemetry data plane workflow")
+
+__Figure 1: Telemetry data plane workflow__
+
+Figure 1 describes the telemetry data plane functionality. The telemetry module inspects every data packet without interfering with normal data forwarding pipeline. A telemetry watchlist table specifies the flows to monitor. It performs ternary match on the packet headers, and provides telemetry action parameters. Packets belonging to the specified flow spaces will be processed by the the event detection logic. If a triggering event is detected, the switch will generate a report message to the monitor. The report message includes packet header and switch metadata associated with the packet (e.g., timestamp, ingress/egress ports, queue latency).
+
+## In-band Network Telemetry ##
+
+![INT](figures/INT.png "Figure 2: In-band Network Telemetry")
+
+__Figure 2: In-band Network Telemetry__
+
+Figure 2 shows an example workflow of In-band Network Telemetry (INT). Switches along the route path add switch metadata into the packet header, based on the “telemetry instructions” carried in the INT header.
+
+Each switch may play the role of __endpoint__ or __transit__ for INT packets. INT endpoint acts both as __INT source__ and __INT sink__. As INT source it initiates INT operation by inserting INT header into a packet and thereby instructing other network devices along the routing path to add desired telemetry information into the packet. As INT sink device it extracts the INT information from the packet and exports it using a mirror session to the monitor.
+
+An INT transit device is a device along the path of a packet from source to sink. INT transit device adds its own INT information to the packet as requested by an INT source.
+
+
+## Packet Postcards ##
+
+![Postcard](figures/postcard.png "Figure 3: Postcard")
+
+__Figure 3: Postcard Telemetry__
+
+Figure 3 shows an example workflow of Packet Postcards, an alternative telemetry approach. Each switch makes its own decision and reports packets info (aka “postcards”) to the monitor. Unlike INT, a Postcard switch never modifies the original data packets. 
+
+## Mirror on Drop ##
+
+![MoD](figures/mod.png "Figure 4: Mirror on Drop")
+
+__Figure 4: Mirror on Drop__
+
+Figure 4 depicts the mirror on drop capability. Switches mirror packets dropped by the ingress pipe, egress pipe or queueing buffer to the monitor for network diagnosis. The report messages include packet header, switch metadata, and drop reason.
```

Comment:
Mirror on drop (MoD) is similar to hostif trap in terms of capturing dropped packets. The main difference is that MoD in this proposal is integrated with the telemetry framework. It unifies all drop conditions, sends dropped packets with drop reasons and associated switch metadata to the (**remote**) monitor through ERSPAN sessions. I am not sure how to nicely integrate it into the existing hostif framework. Please let me know if you have any suggestions. Thanks!